File: src/tools/robot_debug/opengl_gui/widget_model_frame.cpp
#include "widget_model_frame.h"



WidgetModelFrame::WidgetModelFrame()
            :WidgetFrame()
{

}


WidgetModelFrame::WidgetModelFrame(GLVisualisation &visualisation_, Variables &variables_,  LoadTextures &textures_, Json::Value &params_)
            :WidgetFrame(visualisation_, variables_, textures_, params_)
{
  std::string obj_file_name = params["obj_file_name"].asString();
  std::string mtl_file_name = params["mtl_file_name"].asString();
  float scale = params["scale"].asFloat();

  rx = params["rotate"][0].asFloat();
  ry = params["rotate"][1].asFloat();
  rz = params["rotate"][2].asFloat();

  sx = params["shift"][0].asFloat();
  sy = params["shift"][1].asFloat();
  sz = params["shift"][2].asFloat();

  model.init(visualisation_, obj_file_name, mtl_file_name, scale);
}


WidgetModelFrame::WidgetModelFrame(WidgetModelFrame& other)
            :WidgetFrame(other)
{

}

WidgetModelFrame::WidgetModelFrame(const WidgetModelFrame& other)
            :WidgetFrame(other)
{

}

WidgetModelFrame::~WidgetModelFrame()
{

}


void WidgetModelFrame::render()
{
  if (m_visible)
  {
    visualisation->push();
    visualisation->translate(x, y, z);

    render_frame();

    float drx =  variables->v[params["variable"]["name"].asString()][0].asFloat();
    float dry =  variables->v[params["variable"]["name"].asString()][1].asFloat();
    float drz =  variables->v[params["variable"]["name"].asString()][2].asFloat();


    visualisation->push();
      visualisation->translate(sx, sy, sz);
      visualisation->rotate(ry - dry, rx - drx, rz + drz);

      model.render();

    visualisation->pop();

    for (unsigned int i = 0; i < widgets.size(); i++)
      widgets[i]->render();
    visualisation->pop();
  }
}

File: src/tools/robot_debug_py/opengl_gui/widget_model_frame.py
from opengl_gui.widget_frame import WidgetFrame
from opengl_gui.obj_model import ObjModel

class WidgetModelFrame(WidgetFrame):
    def __init__(self, visualisation, variables, textures, params):
        super().__init__(visualisation, variables, textures, params)


        model_file_name = str(params["model file name"])
        
        self.scale = float(params["scale"])

        self.rx = float(params["rotate"][0])
        self.ry = float(params["rotate"][1])
        self.rz = float(params["rotate"][2])

        self.sx = float(params["shift"][0])
        self.sy = float(params["shift"][1])
        self.sz = float(params["shift"][2])

        self.model = ObjModel(model_file_name)

        self.step = 0.0


    def render(self):
        if self.visible:
            self.visualisation.push()
            self.visualisation.translate(self.x, self.y, self.z)

            self.render_frame()

            values = self.variables.v[self.variable_name]

            drx =  float(values[0]) 
            dry =  float(values[1]) + 60
            drz =  float(values[2]) + self.step

            self.visualisation.push()
            self.visualisation.translate(self.sx, self.sy, self.sz)
            self.visualisation.rotate(self.ry - dry, self.rx - drx, self.rz + drz)

            for i in range(len(self.model.polygons)):
                self.visualisation.push()

                color = self.model.colors[i]

                self.visualisation.set_color(color[0], color[1], color[2])
                self.visualisation.paint_polygon(self.model.points, self.model.polygons[i], self.scale)
                self.visualisation.pop()


            self.visualisation.pop()

            for i in range(len(self.child_widgets)):
                self.child_widgets[i].render()
   
            self.visualisation.pop()

            self.step+= 1
